### ISSUE TITLE
clarify training on the input and output vs just training on output

### ISSUE THREAD

**Issue by msalganik** [niznik-dev/predicting-zygosity#17]
I think this is related to masking.
Do we want the model to try to adjust to predict things in the prompt or just the output?

**Comment by msalganik** [comment]
I think we want to be using an instruct dataset https://docs.pytorch.org/torchtune/0.6/basics/instruct_datasets.html#instruct-dataset-usage-label

**Comment by niznik-dev** [comment]
Let's talk more about this at the Software Meeting

**Comment by niznik-dev** [comment]
dataset:
  train_on_input: True

**Comment by niznik-dev** [comment]
False is the default value, so that's what we've been using so far. I'll work on a simple test and see what we can do here.

**Comment by niznik-dev** [comment]
@msalganik Can you add some further thoughts on what we need to merge this one? It sounded like we wanted to try a larger dataset, look into the tokenizer to see if any of the words were multiple tokens, and I'm probably forgetting others.

**Comment by msalganik** [comment]
Sorry that I had to run.  Looping in @DonggyuBan so that he can see how this works since he'll likely be developing tests in the future.

Here's a bit of context on what I think this flag is doing:
- https://github.com/pytorch/torchtune/issues/1534?utm_source=chatgpt.com
- https://chatgpt.com/share/682e554d-cd7c-8012-9f1d-25113de81644 [I'm finding o3 to be AMAZING)

So here's how I'm thinking about an empirical test.

A couple pieces of context:
- Are we tracking runtime (in seconds for example)?  I'd be curious if the runtime between train_on_input: True is the same as train_on_input: False (when ever other parameter is kept constant).
- How is the loss being calculated?  Could you find that part of the code?  Does that include the loss on the input and output when train_on_input is true?  [My guess is that when we train on input that is going into the loss]

Once we know that stuff, one idea I would have is create some input-output pairs where:
- there is a clear pattern in the input, there is a clear relationship between the input and the output
- there is a clear pattern in the input, and no clear relationship between input and output
- there is no pattern in the input, there is no pattern in the output
- there is no pattern in the input, there is a clear relationship between the input and output (I know this is weird but I'll explain)

Here's a way to generate a clear patterned input:
- Pick a random number between 0 and max (say 1000) - k (say 6)
- Let's imagine random number we pick happens to be 42.  Then write the sequence of length k (for now say 6): "42-43-44-45-46-47". This is the clearly patterned input.  The LLM can learn this pattern.  And it is set up in a way that looks clear in "token space" (you can put it in here and see: https://belladoreai.github.io/llama3-tokenizer-js/example-demo/build/) 

Now how do we decide what output goes with this?
If we are in the case where we want a clear relationship you can just put the next token "48".
If we are in the case where we don't want a clear relationship you can just put a random token between 0 and max - k.

Now if we want an unpatterned input:
- Just make k random numbers between 0 and max (I picked 999 because of tokenizer but it probably does not matter).

Now how do we decide what output goes with this?
If we are in the case where we don't want a clear relationship you can just put a random token between 0 and max - k.
If we are in the case where we want a clear relationship you can just put the next token "42".  In this case the model should learn that correct answer is always 42.  You might wonder why 42?  (why not: https://en.wikipedia.org/wiki/Phrases_from_The_Hitchhiker%27s_Guide_to_the_Galaxy#_The_Answer_to_the_Ultimate_Question_of_Life,_the_Universe,_and_Everything_is_42)

Let's say that we want a sample size of n for each of these 4 cases:
- case 1: predictable input, predictable output
- case 2: predictable input, unpredictable output
- case 3: unpredictable input, predictable output
- case 4: unpredictable input, unpredictable output

Then we want to train each one on 
train_on_input: True
train_on_input: False

There would be 8 runs:
- run 1: predictable input, predictable output (train_on_input: True)
- run 2: predictable input, unpredictable output (train_on_input: True)
- run 3: unpredictable input, predictable output (train_on_input: True)
- run 4: unpredictable input, unpredictable output (train_on_input: True)
- run 5: predictable input, predictable output (train_on_input: False)
- run 6: predictable input, unpredictable output (train_on_input: False)
- run 7: unpredictable input, predictable output (train_on_input: False)
- run 8: unpredictable input, unpredictable output (train_on_input: False)

Now time for guesses.  Everyone who wants should add their guesses. 

- I would guess that in the eval set the log loss will go to 0 in runs 1, 3, 5, 7.
- I would guess that in the eval set the log loss will not go to 0 in runs 2, 4, 6, 8.
- If the guess that the training set log loss will be lower in run 1 than run 3 but the training set log loss will be the same in run 5 and run 7.
- I would guess that the reported tokens per second will be lower for runs 5, 6, 7, 8 than runs 1, 2, 3, 4. (because we are not counting the tokens in the input; in fact should probably be able to calculate the difference between the two tokens per second given the number of tokens in the input)
- I would guess that the wall clock time for all 8 runs will be the same.

OK, now a couple of questions for you @niznik-dev: 
- do you think this will help us verify that we understand this flag correctly?
- are there things you want to add?
- do you want to make some guesses?  Everyone else can add guesses too.
- do you think this will be burdensome to do?

To be honest, this might seem like overkill, but I see it as us building our intuition and building our testing skills. 

@niznik-dev if you think a zoom would help, let me know.

**Comment by niznik-dev** [comment]
I think the tests in [55202e4](https://github.com/niznik-dev/predicting-zygosity/pull/25/commits/55202e4606deb4ac8954373ee5e07534edbebd9b) will address what's outlined above, but I won't run them yet because I want to make predictions.

**Comment by niznik-dev** [comment]
Copying Matt's guesses and then adjusting with my thoughts

- eval log loss will go to 0 in runs 1, 5, 7
- eval log loss will not go to 0 in runs 2, **3**, 4, 6, 8

I think that run 3 might not go to 0 because the input will confuse things.

- training log loss will be lower in run 1 than run 3 but the training set log loss will be the same in run 5 and run 7
- I would guess that the reported tokens per second will be lower for runs 5, 6, 7, 8 than runs 1, 2, 3, 4
- I'm expecting train_on_input: True runs to be slightly slower than train_on_input: False runs

**Comment by niznik-dev** [comment]
Results are here: https://wandb.ai/niznik-p-princeton-university/predictable_or_not

**Comment by msalganik** [comment]
Amazing.  How would you summarize these?  Here are some of the patterns that I see and then a two questions:

Here's one way to look at the broad pattern of the results.
![Image](https://github.com/user-attachments/assets/099663de-4c4a-4b54-a39a-7e0b7b4fad60)

Patterns: 
- For all runs with [?p] the val_loss is close to 0.  That is runs 1, 3, 5 & 7.  That's a good sign.
- When [uu]_false (run 8) and [pu]_false (run 6) has the same loss and val_loss.  That's a good sign.
- For all runs with [?u] the val_loss is about the same.  That's a good sign.
- The training loss is going down for everything, and the ordering of loss (in the training data) at the end seems reasonable to me.  It is highest for [uu]_true then [up]_true. [up]_false and [pp]_false are the lowest.  I guess it is an open question about what would happen as the number of epochs goes to infinity.  I think the loss (which is measured on the training data) would go to 0 for all. ![Image](https://github.com/user-attachments/assets/04487b05-3cbc-4aaa-8c58-469cf0e61ca0)


Questions about tokens per second graph:
- Why is the tokens per second so spiky?  And why are there so many different values?  ![Image](https://github.com/user-attachments/assets/cdc6fd08-8520-4026-ae4b-4949b77035fd)

My guess for why it is so spiky has to do with the tokens per second when we are doing training and eval, but I'm not sure.  @niznik-dev do you know?

Also, my guess was that there were was going to be one average tokens per second when train on input was True and other when train on input was False.  Instead it looks like there are maybe 4 groups with a weird pattern.  

One group seems to be [pu]_true, [up]_true, [pp]_true.  These all have about 15,500 tokens per second at the max.  The next group seems to be [uu]_true at about 7800 tokens per second.  I'm not sure why [uu]_true would have about half the tokens per second of the others.  

A third group seems to be [pu]_false, [up]_false, [pp]_false.  These all have about 4500ish tokens per second.  And the last group seems to be [uu]_false at about 2300 tokens per second.  Again, I'm not sure why [uu]_false would have about half the tokens per second that the others (such as [pu]_false).  That's a mystery to me.  

My guess is that something is weird in how the uu books are generated.  @niznik-dev can you take a look at them?  Are they possibly twice as long or something?  Possible supporting evidence for this hypothesis can be seen in the memory use graph.  The [uu]_false and [uu]_true seem to be using way more memory than the other 6 ![Image](https://github.com/user-attachments/assets/003d1c2c-df00-4841-959c-d001c47441d5)

**Comment by msalganik** [comment]
@niznik-dev also for the record I'll score my predictions, and you should do the same.  It is good practice.  And part of writing them up clearly is that others on the team can learn from what we are doing.

- Prediction 1: "I would guess that in the eval set the log loss will go to 0 in runs 1, 3, 5, 7." ✅ 
![Image](https://github.com/user-attachments/assets/e22de284-5e74-4630-8f5f-1b4f7d9c00dd)

- Prediction 2: "I would guess that in the eval set the log loss will not go to 0 in runs 2, 4, 6, 8." ✅ 
Same graph as above.  An even better test would have been for me to predict the actually level of the eval_log loss, which might be possible given how the test in constructed.  @alessandramaranca might have been making progress on these kind of analytical results. 

- Prediction 3: "[I would guess] that the training set log loss will be lower in run 1 than run 3 but the training set log loss will be the same in run 5 and run 7."  🤷 

This prediction is not specific enough to decide because the relatively ordering of the training set log loss varies over time.  

![Image](https://github.com/user-attachments/assets/345e0c7f-be0f-4ada-8362-db3f8f6823da)

But after one epoch we can see that the training loss seems pretty similar for 
- [uu]_false (run 8) and [pu]_false (run 6)
- [pp]_false (run 7) and [up]_false (run 5)

So I guess I was right about run 5 and run 7 being similar.

And the training loss is lower for [pp]_true (run 1) than [up]_true (run 3).  So I guess sort of right about that too.

Lesson: If making guess about training loss, need to specify after how much training.

- Prediction 4: "I would guess that the reported tokens per second will be lower for runs 5, 6, 7, 8 than runs 1, 2, 3, 4. (because we are not counting the tokens in the input; in fact should probably be able to calculate the difference between the two tokens per second given the number of tokens in the input)" ❌ 

It seems like we have four different values for tokens per second not 2.  See post above [https://github.com/niznik-dev/predicting-zygosity/issues/17#issuecomment-2914266429].  I've also suggested a possible reason that this might be the case.

- Prediction 5: "I would guess that the wall clock time for all 8 runs will be the same." ❌ 
![Image](https://github.com/user-attachments/assets/e23d95a9-e409-408e-8b0d-f8cc6b838abb)

For reasons that are not clear to me, it looks like there were 3 different clock times.

Fastest: [pu]_false, [pu]_true, [pp]_false, [pp]_true

Middle: [up]_false, [up]_true

Slowest: [uu]_false, [uu]_true

This is partially consistent with the idea that [uu] is not behaving as expected in terms of timing and tokens per second.

Not sure why [up]_? would be different from [pu]_?.  Perhaps something with how u was done?


Overall not bad, but definitely not perfect.  I'm very curious about the way the tokens per second and run times differed from expectation.

**Comment by niznik-dev** [comment]
Let me do two reruns for uu, it looks like they ended up on a 40 GB A100 so I think that's part of what we're seeing.

**Comment by niznik-dev** [comment]
Done, I would now say we have two clear tokens per second tiers and the times for each run are more standard (I would wager if I redo 3 and 7 they'd be faster - it's probably just I/O hiccups or something).

**Comment by msalganik** [comment]
Great.  For the future, can you put screenshots into the thread to support claims.  That makes it easier for everyone to follow along.

I looked back at W & B and it looks like you've discarded the uu runs and replaced them with new runs.  I now confirm that I see two tokens per second groups as well.

![Image](https://github.com/user-attachments/assets/d8476a85-b139-4f56-8727-a1ba13a99250)

**Comment by niznik-dev** [comment]
Sure, I'll do screenshots more often. I was trying to get this up before my cello lesson last night so it was a bit rushed. 😅

@msalganik I will rerun the long ones to see if this is consistent.

**Comment by msalganik** [comment]
> I was able to see a run 1 that was 2:12 so I think it's just random or based on the node we get. 

Is this related to 40GB vs 80GB?  Could we be more specific?  The reason I ask is that if we see differences that big randomly it is going to be hard for us to test anything.

**Comment by niznik-dev** [comment]
So it does seem like I'm still getting 40 GB GPUs even though I'm requesting only to get 80 GB. Working with the team to see if there's a bug when in the test queue. Good catch!

I'm using Llama-3.2-1B-Instruct but let me see what I can do about the above before we do more.

**Comment by niznik-dev** [comment]
...not a bug, I couldn't tell that I still had the constraint line in my slurm script commented out 🤦🏼

Redoing the tests for speed now.

**Comment by niznik-dev** [comment]
Also would you prefer I leave the "wrong" runs in WandB and upload the corrected ones with new names, or are you OK with what I did previously and just delete the old runs in favor of the corrected ones?

**Comment by niznik-dev** [comment]
For now, I've uploaded the new runs (3 and 7) and renamed the old ones with a "_40G" until I hear about whether deleting is OK. We now have 8 runs with essentially the same runtime:

<img width="1367" alt="Image" src="https://github.com/user-attachments/assets/3947b9b4-ab36-4e04-a718-596ed1dec8d5" />

**Comment by msalganik** [comment]
> Also would you prefer I leave the "wrong" runs in WandB and upload the corrected ones with new names, or are you OK with what I did previously and just delete the old runs in favor of the corrected ones?

To be honest, I'm not sure.  I think it depends on context.  Since these are all cheap to generate, I don't think it matters to much.  For days that have multiple runs, let's not delete them.

**Comment by niznik-dev** [comment]
I deleted the old runs since these take 2 minutes to run. We should be good on the 8 run comparison now.

**Comment by msalganik** [comment]
@niznik-dev I see no reason not to merge this in if you think it is ready.

Also, as we discussed at the meeting, I think this creates some new questions about spiky tokens per second, but that's a "performance issue" not a "correctness issue".

Feel free to merge and close when you think ready.
